File: LeetCode/Medium/2257-count-unguarded-cells-in-the-grid/2257-count-unguarded-cells-in-the-grid.js
/**
 * @param {number} m
 * @param {number} n
 * @param {number[][]} guards
 * @param {number[][]} walls
 * @return {number}
 */
var countUnguarded = function(m, n, guards, walls) {
    let mat = new Array(m);
    let set = new Set();
    let count = 0;

    for (let i = 0; i < m; i++) {
        mat[i] = new Array(n);
    }

    for (let i = 0; i < walls.length; i++) {
        let key = JSON.stringify(walls[i]);
        set.add(key);
    }

    for (let i = 0; i < guards.length; i++) {
        let [r,c] = guards[i];
        while (c < n) {
            if (set.has(JSON.stringify([r,c]))) {
                break;
            }
            mat[r][c] = "g";
            c++;
        }
        c = guards[i][1];
        while (c >= 0) {
            if (set.has(JSON.stringify([r,c]))) {
                break;
            }
            mat[r][c] = "g";
            c--;
        }
        c = guards[i][1];
        while (r < m) {
            if (set.has(JSON.stringify([r,c]))) {
                break;
            }
            mat[r][c] = "g";
            r++;
        }
        r = guards[i][0];
        while (r >= 0) {
            if (set.has(JSON.stringify([r,c]))) {
                break;
            }
            mat[r][c] = "g";
            r--;
        }
    }
    for (let i = 0; i < m; i++){
        for (let j = 0; j < n; j++) {
            if (!mat[i][j] && !set.has(JSON.stringify([i,j]))) {
                count ++;
            }
        }
    }
    return count;
};

File: LeetCode/Medium/2257-count-unguarded-cells-in-the-grid/2257-count-unguarded-cells-in-the-grid.py
class Solution:
    def countUnguarded(self, m: int, n: int, guards: List[List[int]], walls: List[List[int]]) -> int:
        mat = [[0]*n for i in range(m)]
        
        for guard in guards:
            r, c = guard
            mat[r][c] = "G"
        
        for wall in walls:
            r,c = wall
            mat[r][c] = "W"
        
        path = {"l":[0,-1], "r":[0,1], "u":[-1,0], "d":[1,0]}

        def walk(r,c,p):
            nonlocal m,n
            if r in {-1, m} or c in {-1, n} or mat[r][c] in {"G", "W"}:
                return
            mat[r][c] = "g"
            walk(r + path[p][0], c + path[p][1], p)

        for guard in guards:
            r,c = guard
            walk(r-1,c,"u")
            walk(r+1,c,"d")
            walk(r,c-1,"l")
            walk(r,c+1,"r")
           
        count = 0
        for i in range(m):
            for j in range(n):
                if mat[i][j] == 0:
                    count += 1

        return count